Source/NSFontCollection.m

```diff
@@ -0,0 +1,341 @@
+  RELEASE(_queryAttributes);
+  [super dealloc];
+}
+
+// This method will get the actual list of fonts
+- (void) _runQueryWithDescriptors: (NSArray *)queryDescriptors
+{
+  NSEnumerator *en = [queryDescriptors objectEnumerator];
+  GSFontEnumerator *fen = [GSFontEnumerator sharedEnumerator];
+  id d = nil;
+
+  ASSIGNCOPY(_queryDescriptors, queryDescriptors);
+  while ((d = [en nextObject]) != nil)
+    {
+      NSArray *names = [fen availableFontNamesMatchingFontDescriptor: d];
+      id name = nil;
+
+      en = [names objectEnumerator];
+      while ((name = [en nextObject]) != nil)
+        {
+          NSFont *font = [NSFont fontWithName: name size: 0.0]; // get default size
+          [_fonts addObject: font];
+        }
+    }
+}
+
++ (NSFontCollection *) fontCollectionWithDescriptors: (NSArray *)queryDescriptors
+{
+  NSFontCollection *fc = [[NSFontCollection alloc] init];
+  ASSIGNCOPY(fc->_queryDescriptors, queryDescriptors);
+  return fc;
+}
+
++ (NSFontCollection *) fontCollectionWithAllAvailableDescriptors
+{
+  return [self fontCollectionWithDescriptors:
+                 [[GSFontEnumerator sharedEnumerator] availableFontDescriptors]];
+}
+
++ (NSFontCollection *) fontCollectionWithLocale: (NSLocale *)locale
+{
+  return [self fontCollectionWithAllAvailableDescriptors];
+}
+
++ (BOOL) showFontCollection: (NSFontCollection *)collection
+                   withName: (NSFontCollectionName)name
+                 visibility: (NSFontCollectionVisibility)visibility
+                      error: (NSError **)error
+{
+  NSNumber *v = [NSNumber numberWithInt: visibility];
+  [__sharedFontCollections setObject: collection
+                             forKey: name];
+  [__sharedFontCollectionsVisibility setObject: v
+                                        forKey: name];
+  return YES;
+}
+
++ (BOOL) hideFontCollectionWithName: (NSFontCollectionName)name
+                         visibility: (NSFontCollectionVisibility)visibility
+                              error: (NSError **)error
+{
+  
+  [__sharedFontCollectionsHidden addObject: name];
+  return YES;
+}
+
++ (BOOL) renameFontCollectionWithName: (NSFontCollectionName)aname
+                           visibility: (NSFontCollectionVisibility)visibility
+                               toName: (NSFontCollectionName)name
+                                error: (NSError **)error
+{
+  NSFontCollection *fc = [__sharedFontCollections objectForKey: aname];
+  [__sharedFontCollections setObject: fc forKey: name];
+  return YES;
+}
+
++ (NSArray *) allFontCollectionNames
+{
+  return [__sharedFontCollections allKeys];
+}
+
++ (NSFontCollection *) fontCollectionWithName: (NSFontCollectionName)name
+{
+  return [__sharedFontCollections objectForKey: name];
+}
+
++ (NSFontCollection *) fontCollectionWithName: (NSFontCollectionName)name
+                                   visibility: (NSFontCollectionVisibility)visibility
+{
+  return [__sharedFontCollections objectForKey: name];
```

Comment:
This looks wrong. It seems we don't quite understand the visibility concept here. This seems to be independent of show/hide.